tests/lang/s02_core/s02_inline_iterators/t02_inline_iterators.nim

```diff
@@ -0,0 +1,169 @@
+## elements to, and then assert they were pushed in the right order.
+## For example - double yield with elements added in the loop body
+## itself.
+block order_assertion_example:
+  var values: seq[string]
+
+  iterator orderExample(): int = 
+    ## Iterator can access variables outside of the loop just as
+    ## easily as a regular procedure.
+    values.add "started iterator"
+    yield 12
+    values.add "after yield"
+    yield 24
+    values.add "another yield"
+
+  for value in orderExample():
+    values.add "got " & $value & " from the iterator"
+
+  doAssert values == @[
+    "started iterator",
+    "got 12 from the iterator",
+    "after yield",
+    "got 24 from the iterator",
+    "another yield"
+  ]
+
+block iterator_state_preservation:
+  ## When iterator is executed it's whole state is preserved 
+  ## even when control flow is transferred to the loop body.
+  var values: seq[string]
+
+  iterator storeState(): int = 
+    var state = "starting state"
+    yield 1
+    values.add state
+    
+    ## Changing internal state of the iterator here - "after yield"
+    ## should be placed in values after the element from the loop
+    ## body.
+    state = "after yield"
+    yield 2
+    values.add state
+
+  for item in storeState():
+    values.add "got " & $item
+
+  doAssert values == @[
+    "got 1",
+    "starting state",
+    "got 2",
+    "after yield"
+  ]
+
+  ## As you can see, list of values contains result of the first 
+  ## yield, original state, then second yield and value of the changed 
+  ## state.
+
+## It is still possible to use loop control flow constructs - `break` and
+## `continue`.
+block iterator_and_break:
+  block clean_break:
+    ## When break occurs in the for loop control no longer returns to
+    ## the iterator body.
+    var values: seq[string]
+    iterator breakIn(): int = 
+      values.add "before 1"
+      yield 1
+      values.add "after 1"
+
+      values.add "before 2"
+      yield 2
+
+      ## If `break` was called while result of the `yield 2` was being 
+      ## processed this statement won't be executed.
+      values.add "after 2"
+
+    for value in breakIn():
+      values.add $value
+      if value == 2:
+        break 
+
+    doAssert values == @[
+      "before 1",
+      "1",
+      "after 1",
+      "before 2",
+      "2"
+    ]
+
+  block break_try_finally:
+    ## If iterator maintains an internal state that have to be properly cleaned up
+    ## `try: ... finally:` block can be used to ensure that resources are 
+    ## properly dispoed of.
+
+    var values: seq[string] 
+    iterator breakFinally(): int = 
+      values.add "resource setup"
+
+      try:
+        values.add "before 1"
+        yield 1
+        ## If break is called control flow won't return to this part 
+        ## of the code 
+        values.add "after 1"
+
+      finally:
+        values.add "resource cleanup"
+
+    for item in breakFinally():
```

Comment:
Todo, noted for later, need to add in the rules for loop items/pairs inference.

It should actually be changed, like either static interfaces, rows, concepts, or something to make it work nicely.